Camp crafting uses the normal crafting GUI

RenechCDDA opened the pull request:
#### Summary
Interface "Camp crafting uses the normal crafting GUI"

#### Purpose of change
Camp crafting is problematic in many ways. Accessing the recipes, manipulating them, seeing if they're available, etc.

![image](https://github.com/user-attachments/assets/e39ac973-8693-4c54-b80a-65e179398573)

We have a perfectly good crafting GUI already, why not use it?

#### Describe the solution
Use the existing crafting GUI.

We shove all the recipes and proficiencies from our camp's NPCs into a dummy, give them access to the camp inventory, and put them into the crafting window. We wait for the return value and instead of doing a normal crafting activity we craft like camps always used to, by off-mapping a chosen NPC.

Essentially we are borrowing the crafting GUI, with a few little changes to make it work for us.

#### Describe alternatives you've considered
Making the mission window searchable/filterable :^)

#### Testing
**DRAFT**


https://github.com/user-attachments/assets/c2d5def9-3153-4bba-ae34-f9a2d70ec967



The draft as posted does not work ~~(inventory is not properly tallied, somehow) and~~ it does not do anything with the recipe selected in the crafting window.

#### Additional context

Known issues:
The flow of recipe selection --> crafter selection is maybe not ideal

Can't see kcal display until after you select a recipe

Descriptions for camp-provided recipes are still loaded and sent for translation, but we don't use them

RenechCDDA committed:
Basic implementation of crafting GUI accessed thru camp

PatrikLundell commented:
Looks like a very intriguing idea, and I really hope it's successful. It would be a major improvement to base camp crafting.

And now comes the downer section with issues that will have to be dealt with at the tail end (view this as a reminder, rather than mistrust of the OP's abilities):
- There's a small number of recipes available to companions only for base camps. As far as I know they all use the drop hammer built in some base camps and in the workshop expansions, but there may be additional ones. They'd need to be dealt with in some way, with the easiest one probably being to make them available to everyone (i.e. not companion only), and make them available through either skill levels or a book.
- Make all functionality for the base camp (including expansions) available, not only regular tools and material. I think that would involve pseudo tools only, but I may be wrong, and it may already have been dealt with.
- Get rid of all the recipe couplings of base camps and expansions as they're no longer used. That's just a cleanup thing, but there may be a need to review construction mission texts and possibly base camp setup as well, to get rid of obsolete references, so it might be a bit more work than it seems initially.

RenechCDDA commented:
My overall plan at this time was to move basecamp-specific recipes to their own crafting category, since otherwise they'll be very difficult to find.

I don't see a need to make (e.g.) drop hammer recipes generally available, they can still be provided by the basecamp expansions. If anything I wanted to expand on that - the expansions being able to provide access to recipes is a *good* thing.

One thing currently missing is the calorie cost display for crafting, I don't know how to deal with that. Maybe as a final query before you the mission starts.

PatrikLundell commented:
I disagree with you regarding expansion specific recipes, as my preference is to make bases and expansions optional, but I don't disagree violently with it.

If you've got trouble with calories you've probably also go problems with time, as the time given for normal crafting is work time, while the base camp craft time is given in calendar time, and if you've figured out how to intervene in the display of things to show one you've probably done most of the work for the other, as calendar time is derived from work time and calories from calendar time and exercise level.
Also note that both of these are modified by the number of people involved (as the number of time off periods can change a by up to one for each additional person assigned).

RenechCDDA committed:
-Override normal companion selection
-Popups to inform about kcal costs/query to exit if you balked at the cost
-Remove some now-unused arguments
-Cleanup comments
-Clang BE SILENCED
-Fix batch crafting

RenechCDDA commented:
This is mostly working now, except batch selection is broken. I think I just missed a call to `availability` so that should be easy to squash.

Translations I think will be dealt with by simply removing them.

Not aware of any other huge issues, but there may be problems found once it gets some general testing.

RenechCDDA commented:
Ehhh translations look like it won't be so simple because we *want* to keep translations for the expansions, but discard them only for recipes. Both are loaded as `recipe_group`, so it's not simple.

I guess I can kick that can down the road (to maybe someone else) and mark this ready for review, huh? Let's see if it blows up.

RenechCDDA committed:
Various clang fixes

Night-Pryanik commented on src/faction_camp.cpp:
```diff
@@ -3512,62 +3508,113 @@ void basecamp::start_combat_mission( const mission_id &miss_id, float exertion_l
-            return;
+    validate_sort_points();
+    form_crafting_inventory( get_camp_map() );
+
```
```suggestion
```

RenechCDDA replied:
I'm having trouble seeing what this suggestion is, github's diff highlighting is not making it clear. You're suggesting I remove the sort point validation and formation of the crafting inventory before we call select_crafter_and_crafting_recipe?

Night-Pryanik replied:
That's VERY weird and obviously not what I had in mind. GitHub sometimes confuses me. Let's close this suggestion.

Zireael07 commented:
Finally!

Brambor commented on src/crafting_gui.cpp:
```diff
@@ -1344,7 +1348,10 @@ std::pair<Character *, const recipe *> select_crafter_and_crafting_recipe( int &
 
     // Get everyone's recipes
-    const recipe_subset &available_recipes = crafter->get_group_available_recipes();
+    // WTF? If called with dummy npc, we have to do this. Why? Why doesn't Character::get_group_available_recipes()
```
Looking at the implementation, Does the dummy have the inventory accessible through `dummy->crafting_inventory()` ? I think it is very separate in `inventory_override`.

RenechCDDA replied:
Should be resolved with #81148

Brambor commented on src/faction_camp.cpp:
```diff
@@ -1231,12 +1227,13 @@ void basecamp::get_available_missions_by_dir( mission_data &mission_key, const p
     {
         mission_id miss_id = { Camp_Crafting, "", {}, dir };
         comp_list npc_list = get_mission_workers( miss_id, true );
+        // WTF?
```
Not to be pedantic, but I don't know what caused the `WTF`.

Brambor reviewed (commented):
I know. Merged PR.

 - I am thinking of doing a followup PR if I manage to make the npc selection in choose crafter #68183
 - Also, show crafter and calories here somewhere
   - Maybe select default crafter first so that the player isn't surprised why upon confirmation X started crafting it when they didn't choose anyone (X was choosen automatically all along and they didn't change the crafter).
 -  ![image](https://github.com/user-attachments/assets/3e9ba531-9f68-4453-86cc-13c0a3ae564e)
